File: 20-valid-parentheses/cpp/src/main.cpp
/* ========================================================================
 *
 *     Filename:  
 *  Description:  
 *       Author:  
 *      Version:  0.0.1
 *
 * ======================================================================== */
#include <iostream>
#include <string>
#include <stack>

auto main() -> int
{
	std::cout << "Hello world!\n";
	return 0;
}

class Solution {
public:
    bool isValid(std::string s) {
	auto st = std::stack<char>();

	for (const auto& ch : s) {
		char check = '\0';
		switch (ch) {
			case '(':
			case '{':
			case '[':
				st.push(ch);
				continue;
			case ')':
				check = '(';
				break;
			case '}':
				check = '{';
				break;
			case ']':
				check = '[';
				break;
			default: continue;
		}
		if (not st.empty()) {
			if (st.top() == check)
				st.pop();
			else
				return false;
		} else {
			return false;
		}
			

	}

	return st.empty();
    }
};

File: 20-valid-parentheses/rust/src/main.rs
fn main() {
    println!("Hello, world!");
}

impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut chvec = Vec::new();
        #[derive(PartialEq)]
        enum Parens {
            Clamp,
            Parenthesis,
            Bracket,
        }
        for ch in s.chars() {
            match ch {
                '(' => chvec.push(Parens::Parenthesis),
                '{' => chvec.push(Parens::Bracket),
                '[' => chvec.push(Parens::Clamp),
                ')' => if chvec.pop() == Some(Parens::Parenthesis) { continue } else { return false; },
                '}' => if chvec.pop() == Some(Parens::Bracket) { continue } else { return false; },
                ']' => if chvec.pop() == Some(Parens::Clamp) { continue } else { return false; },
                _ => unreachable!(),
            }
        }

        return chvec.is_empty();
    }
}
